src/types/composite/container.ts

```diff
@@ -1,36 +1,64 @@
+import {ContainerByteArrayHandler, ContainerStructuralHandler, ContainerTreeHandler} from "../../backings";
 import {ObjectLike} from "../../interface";
-import {CompositeType} from "./abstract";
+import {GIndexPathKeys, GINDEX_LEN_PATH} from "../../util/gIndex";
+import {getPowerOfTwoCeil} from "../../util/math";
 import {isTypeOf} from "../basic";
 import {Type} from "../type";
-import {ContainerStructuralHandler, ContainerTreeHandler, ContainerByteArrayHandler} from "../../backings";
+import {CompositeType} from "./abstract";
 
-export interface IContainerOptions {
+export interface IContainerOptions<T extends ObjectLike = ObjectLike> {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  fields: Record<string, Type<any>>;
+  fields: Record<keyof T, Type<any>>;
 }
 
 export const CONTAINER_TYPE = Symbol.for("ssz/ContainerType");
 
 export function isContainerType<T extends ObjectLike = ObjectLike>(type: unknown): type is ContainerType<T> {
   return isTypeOf(type, CONTAINER_TYPE);
 }
-
 export class ContainerType<T extends ObjectLike = ObjectLike> extends CompositeType<T> {
   // ES6 ensures key order is chronological
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  fields: Record<string, Type<any>>;
-  constructor(options: IContainerOptions) {
+  fields: Record<keyof T, Type<any>>;
+  constructor(options: IContainerOptions<T>) {
```

Comment:
Can you split out these type updates to a separate PR?
This can be merged pretty quickly

Reply:
Sure, I'm gonna try to introduce abstract and some more codesharing between basic and composite array and test against lodestar